Feature/allow older db

MikeNeilson opened the pull request:
Fixes #409

MikeNeilson committed:
Initial centralization of auth logic.

MikeNeilson committed:
Add condition logic for multi office auth.

MikeNeilson committed:
Spelling.

MikeNeilson commented:
FYI, after review it might seem rather obvious that "umm, we really only need one access manager now" and yes, but that work will come later to shift the different auth logic to a before handler or something.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/security/OpenIDAccessManager.java:
```diff
@@ -68,48 +66,22 @@ public OpenIDAccessManager(String wellKnownUrl, String issuer, int realmKeyTimeo
-    }
-
-    private String getUserFromToken(Context ctx) {
+    private Optional<DataApiPrincipal> getUserFromToken(Context ctx) throws CwmsAuthException {
```
return is never an empty Optional. Should just return DataApiPrincipal instead

MikeNeilson replied:
Yeah, I mostly just did that to copy the structure of others. I plan to rearrange all this later to pass the generated DataApiPrincipal to a single AccessManager so all of this should be considered a stopgap.

adamkorynta replied:
Sounds good, I got worried about the unchecked `Optional.get()` but then saw there was no nullable `Optional`

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -0,0 +1,223 @@
+     * @param p
+     * @param routeRoles
+     */
+    public static void isAuthorized(Context ctx, Optional<DataApiPrincipal> p, Set<RouteRole> routeRoles) throws CwmsAuthException {    
```
Anti-pattern using an Optional parameter

MikeNeilson replied:
... really? seems better than a null check.

adamkorynta replied:
Nulls, Optionals , and boolean flag parameters all fall under the same antipattern. It winds up having the function caller needing to know that there are essentially different implementations of the same method that route depending on which parameters were used. This becomes a maintenance issue in public API's as you then have the function callers dictating the implementation support rather than having the API's dictate what they support. 

https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3

MikeNeilson replied:
Okay, I can buy that. Having an actual function name could often be more readable.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -0,0 +1,223 @@
+                    }
+                }
+            } catch (SQLException ex) {
+                logger.atWarning().withCause(ex).log("Failed API key check");
```
The ApiServlet class already logs all thrown exceptions, the log messages in this method don't add any meaningful context.

MikeNeilson replied:
Hmm, yeah but it doesn't have the details. I'll rework the exception too.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/data/dao/AuthDao.java:
```diff
@@ -0,0 +1,223 @@
+     * @throws SQLException
+     */
+    private void setSessionForAuthCheck(Connection conn) throws SQLException {
+        if (hasEnvFix) {
```
Worth some comments describing what hasEnvFix means.

MikeNeilson replied:
good point.

adamkorynta commented on cwms-data-api/src/main/java/cwms/cda/security/CwmsAccessManager.java:
```diff
@@ -37,78 +27,19 @@ public class CwmsAccessManager extends CdaAccessManager {
-        }
+        Optional<DataApiPrincipal> p = getApiPrincipal(ctx);
+        AuthDao.isAuthorized(ctx,p,requiredRoles);
+        AuthDao.prepareContextWithUser(ctx, p.get());
```
Calling `get()` on Optional without isPresent()

MikeNeilson committed:
Incorporate PR feedback.

Remove Optional where it clearly wasn't needed.
Reorganized ApiServlet CwmsAuthException handling to log details but report generic message.
Some general cleanup of the code flow.

MikeNeilson committed:
Update variable name for clarity.
